File: phone/acg-android/app/src/main/java/com/aicraftspeopleguild/acg/data/SyncEngine.kt
package com.aicraftspeopleguild.acg.data

import kotlinx.serialization.json.Json
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext

/**
 * Pulls the GitHub-Issues tag DB into the local [TagDb], re-publishing
 * every fetched tag over [Mqtt] so live subscribers update in place.
 *
 * One-way-for-now: GH Issues -> device. Outbound writes still go direct
 * through [GhTagClient.writeTag]; after a successful write we refresh
 * the one affected path in the local cache.
 */
class SyncEngine(
    private val tagDb: TagDb,
    private val ghTag: GhTagClient,
) {
    private val json = Json { ignoreUnknownKeys = true }

    data class Result(val tagsPulled: Int, val errors: Int, val durationMs: Long)

    suspend fun pullAllTags(): Result = withContext(Dispatchers.IO) {
        val start = System.currentTimeMillis()
        var ok = 0; var errs = 0
        runCatching { ghTag.listTags() }.getOrDefault(emptyList()).forEach { issue ->
            val path = issue.title.removePrefix("tag:")
            runCatching { parseBody(issue.body) }.getOrNull()?.let { v ->
                tagDb.upsertTag(path, v, issue.number)
                Mqtt.publish("tag.$path", v.value, retained = true, from = "sync")
                ok++
            } ?: run { errs++ }
        }
        tagDb.setMeta("last_full_sync", System.currentTimeMillis().toString())
        Result(ok, errs, System.currentTimeMillis() - start)
    }

    /** Write to GH, then refresh just that one path locally. */
    suspend fun writeTagAndSync(
        path: String, value: String, type: String = "String",
        quality: String = "good", description: String = "written via ACG Android",
    ): String = withContext(Dispatchers.IO) {
        val outcome = ghTag.writeTag(path, value, type, quality, description)
        runCatching { ghTag.readTag(path) }.getOrNull()?.let { v ->
            tagDb.upsertTag(path, v, issueNumber = null)
            Mqtt.publish("tag.$path", v.value, retained = true, from = "write")
        }
        outcome
    }

    /** Refresh one tag from GH into the local DB (without writing). */
    suspend fun pullTag(path: String): TagValue? = withContext(Dispatchers.IO) {
        val v = runCatching { ghTag.readTag(path) }.getOrNull() ?: return@withContext null
        tagDb.upsertTag(path, v, issueNumber = null)
        Mqtt.publish("tag.$path", v.value, retained = true, from = "pull")
        v
    }

    private fun parseBody(body: String?): TagValue {
        if (body.isNullOrBlank()) return TagValue()
        var t = body.trim()
        if (t.startsWith("```")) {
            t = t.substringAfter("\n", t)
            if (t.endsWith("```")) t = t.substring(0, t.length - 3)
            if (t.trimStart().startsWith("json")) t = t.trimStart().removePrefix("json").trimStart()
        }
        return try { json.decodeFromString(TagValue.serializer(), t) } catch (_: Exception) { TagValue() }
    }
}

File: phone/acg-ios/ACG/Data/SyncEngine.swift
import Foundation

/// Pulls GitHub-Issues tag DB into the local TagDb and fans each
/// fetched tag out on the Mqtt bus.
final class SyncEngine {
    let tagDb: TagDb
    let ghTag: GhTagClient

    init(tagDb: TagDb, ghTag: GhTagClient) {
        self.tagDb = tagDb
        self.ghTag = ghTag
    }

    struct Result { let pulled: Int; let errors: Int; let durationMs: Int }

    func pullAllTags() async -> Result {
        let start = Date()
        var ok = 0, errs = 0
        do {
            let issues = try await ghTag.listTags()
            for issue in issues {
                let path = String(issue.title.dropFirst("tag:".count))
                if let tv = parseBody(issue.body) {
                    tagDb.upsert(path: path, value: tv, issueNumber: issue.number)
                    Mqtt.shared.publish(topic: "tag.\(path)", value: tv.value ?? "", retained: true, from: "sync")
                    ok += 1
                } else { errs += 1 }
            }
        } catch {
            errs += 1
        }
        tagDb.setMeta("last_full_sync", "\(Int(Date().timeIntervalSince1970 * 1000))")
        return Result(pulled: ok, errors: errs, durationMs: Int(Date().timeIntervalSince(start) * 1000))
    }

    func writeAndSync(_ path: String, value: String, type: String = "String",
                      quality: String = "good", description: String = "written via ACG iOS") async throws -> String {
        let outcome = try await ghTag.writeTag(path, value: value, type: type,
                                               quality: quality, description: description)
        if let tv = try? await ghTag.readTag(path) {
            tagDb.upsert(path: path, value: tv, issueNumber: nil)
            Mqtt.shared.publish(topic: "tag.\(path)", value: tv.value ?? "", retained: true, from: "write")
        }
        return outcome
    }

    private func parseBody(_ body: String?) -> GhTagClient.TagValue? {
        guard var t = body?.trimmingCharacters(in: .whitespacesAndNewlines), !t.isEmpty else { return nil }
        if t.hasPrefix("```") {
            if let idx = t.firstIndex(of: "\n") { t = String(t[t.index(after: idx)...]) }
            if t.hasSuffix("```") { t.removeLast(3) }
            if t.hasPrefix("json") { t.removeFirst(4); t = t.trimmingCharacters(in: .whitespaces) }
        }
        return try? JSONDecoder().decode(GhTagClient.TagValue.self, from: Data(t.utf8))
    }
}
